src/probe_modules/module_dns.c

```diff
@@ -0,0 +1,1043 @@
+/*
+ * ZMap Copyright 2015 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+/* Module for scanning for open UDP DNS resolvers. 
+ *
+ * This module optionally takes in an argument of the form "TYPE,QUESTION"
+ * (e.g. "A,google.com").  
+ *
+ * Given no arguments it will default to asking for an A record for www.google.com.
+ * 
+ * This module does minimal answer verification. It only verifies that the 
+ * response roughly looks like a DNS response. It will not, for example,
+ * require the QR bit be set to 1. All such analysis should happen offline.
+ * Specifically, to be included in the output it requires:
+ * - That the response packet is >= the query packet.
+ * - That the ports match and the packet is complete.
+ * To be marked as success it also requires:
+ * - That the response bytes that should be the ID field matches the send bytes.
+ * - That the response bytes that should be question match send bytes.
+ * To be marked as app_success it also requires:
+ * - That the QR bit be 1 and rcode ==0.
+ *
+ * Usage: zmap -p 53 --probe-module=dns --probe-args="ANY,www.example.com" -O json --output-fields=* 8.8.8.8
+ * 
+ * Based on a deprecated udp_dns module. 
+ */ 
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <string.h>
+#include <assert.h>
+
+#include "../../lib/includes.h"
+#include "../../lib/random.h"
+#include "../../lib/xalloc.h"
+#include "probe_modules.h"
+#include "packet.h"
+#include "logger.h"
+#include "module_dns.h"
+#include "module_udp.h"
+#include "../fieldset.h"
+
+#define DNS_SEND_LEN 512    // This is arbitrary 
+#define UDP_HEADER_LEN 8
+#define PCAP_SNAPLEN 3000   // This is even more arbitrary
```

Comment:
After off-list discussion moving to 1500. (Note, rrsig's for google, for example, are upwards of 1000 bytes.)